aleph-client/src/utility.rs

```diff
@@ -8,64 +8,63 @@ use crate::{
-    async fn get_block_hash(&self, block: BlockNumber) -> Option<BlockHash>;
-    async fn get_best_block(&self) -> BlockNumber;
-    async fn get_finalized_block_hash(&self) -> BlockHash;
-    async fn get_block_number(&self, block: BlockHash) -> Option<BlockNumber>;
+    async fn first_block_of_session(
+        &self,
+        session: SessionIndex,
+    ) -> anyhow::Result<Option<BlockHash>>;
+    async fn get_block_hash(&self, block: BlockNumber) -> anyhow::Result<Option<BlockHash>>;
+    async fn get_best_block(&self) -> anyhow::Result<Option<BlockNumber>>;
+    async fn get_finalized_block_hash(&self) -> anyhow::Result<BlockHash>;
+    async fn get_block_number(&self, block: BlockHash) -> anyhow::Result<Option<BlockNumber>>;
 }
 
 #[async_trait::async_trait]
 pub trait SessionEraApi {
-    async fn get_active_era_for_session(&self, session: SessionIndex) -> EraIndex;
+    async fn get_active_era_for_session(&self, session: SessionIndex) -> anyhow::Result<EraIndex>;
 }
 
 #[async_trait::async_trait]
 impl BlocksApi for Connection {
-    async fn first_block_of_session(&self, session: SessionIndex) -> Option<BlockHash> {
-        let period = self.get_session_period().await;
+    async fn first_block_of_session(
+        &self,
+        session: SessionIndex,
+    ) -> anyhow::Result<Option<BlockHash>> {
+        let period = self.get_session_period().await?;
         let block_num = period * session;
 
         self.get_block_hash(block_num).await
     }
 
-    async fn get_block_hash(&self, block: BlockNumber) -> Option<BlockHash> {
+    async fn get_block_hash(&self, block: BlockNumber) -> anyhow::Result<Option<BlockHash>> {
         info!(target: "aleph-client", "querying block hash for number #{}", block);
         self.client
             .rpc()
             .block_hash(Some(block.into()))
             .await
-            .unwrap()
+            .map_err(|e| e.into())
     }
 
-    async fn get_best_block(&self) -> BlockNumber {
-        self.client
-            .rpc()
-            .header(None)
-            .await
-            .unwrap()
-            .unwrap()
-            .number
+    async fn get_best_block(&self) -> anyhow::Result<Option<BlockNumber>> {
+        self.get_block_number_inner(None).await
     }
 
-    async fn get_finalized_block_hash(&self) -> BlockHash {
-        self.client.rpc().finalized_head().await.unwrap()
-    }
-
-    async fn get_block_number(&self, block: BlockHash) -> Option<BlockNumber> {
+    async fn get_finalized_block_hash(&self) -> anyhow::Result<BlockHash> {
         self.client
             .rpc()
-            .header(Some(block))
+            .finalized_head()
             .await
-            .unwrap()
-            .map(|h| h.number)
+            .map_err(|e| e.into())
+    }
+
+    async fn get_block_number(&self, block: BlockHash) -> anyhow::Result<Option<BlockNumber>> {
+        self.get_block_number_inner(Some(block)).await
```

Comment:
if you convert `get_block_number_inner` into a private method, I would move it to this module

Reply:
done